### PR TITLE
Introducing braketest & operation_mode support

### PR DESCRIPTION
**Pull request by ct2034**

### CODE REVIEW

File: prbt_default_application/scripts/myApplication.py [hunk 1]
```diff
@@ -23,8 +23,8 @@ def start_program():
     r.move(Ptp(goal=start_pos, vel_scale=__ROBOT_VELOCITY__))
 
     rospy.loginfo("Start loop") # log
-    while(True):
-        # do infinite loop
+    for _ in range(5):
```
Comment (jschleicher) [inline prbt_default_application/scripts/myApplication.py]: Why? An application should run as long as the application runs. In C++ the usual use is `while(ros::ok()), in python it's a pure `while True:`.
Inside the loop, the user would e.g. wait, unitl parts from a feeder are available.

**Review by jschleicher — CHANGES_REQUESTED** [review]
Thank you. Please see inline comment.

File: prbt_default_application/scripts/myApplication.py (hunk 1 above)
Comment (ct2034) [reply prbt_default_application/scripts/myApplication.py]: maybe we can do something like https://wiki.ros.org/rospy/Overview/Initialization%20and%20Shutdown#Testing_for_shutdown

Comment (jschleicher) [reply prbt_default_application/scripts/myApplication.py]: Yep, using rospy.is_shutdown() sounds good.
```suggestion
    while not rospy.is_shutdown():
        # do infinite loop
```

File: prbt_default_application/launch/my_application.launch [hunk 2]
```diff
@@ -59,9 +59,15 @@ limitations under the License.
 
       <!-- Safe-torque-off-input from the safety controller is read to brake the robot before power-off
            The signal is read via modbus and you can configure the modbus server ip -->
-      <!-- <arg name="sto" value="pnoz" /> -->
+      <!-- <arg name="sto" value="pss4000" /> -->
       <!-- <arg name="sto_modbus_server_ip" value="192.168.0.10" /> -->
 
+      <!-- You can additionally select the two safety functions below: -->
+      <!-- Set braketest usage -->
+      <arg name="has_braketest_support" default="true"/>
```
Comment (jschleicher) [inline prbt_default_application/launch/my_application.launch]: Are braketest and operation mode enabled on purpose? "You can additionally" sounds like you would let the user choose, if he wants to enable this feature? Selectable features are commented in the template to let the user change the default value by uncommenting the line(s).
```suggestion
      <!-- <arg name="has_braketest_support" default="true"/> -->
```

**Review by agutenkunst — APPROVED** [review]

**Review by jschleicher — COMMENTED** [review]
Please see the inline comment. 
Overall :+1:

File: prbt_default_application/launch/my_application.launch (hunk 2 above)
Comment (ct2034) [reply prbt_default_application/launch/my_application.launch]: Yes. But I think the whole setup including control cabinet etc is going to be our default setup, right? So I change the wording in the comment, ok?

Comment (jschleicher) [reply prbt_default_application/launch/my_application.launch]: Or change the default to true in robot.launch (if that setup is the main use case) and leave a commented `false` in here to disable it...

Comment (jschleicher) [reply prbt_default_application/launch/my_application.launch]: @ct2034 As you say braketest support is the default use case. So I'd still suggest to change the default to `true` in [robot.launch](https://github.com/PilzDE/pilz_robots/blob/2c0a3060b2670c3be2aeea19c7316365e53d7f28/prbt_support/launch/robot.launch#L18) and leave a commented `false` here in the templates to disable it if needed.

**Review by jschleicher — CHANGES_REQUESTED** [review]
As said before: `has_braketest_support` should probably be set to true in robot.launch.

**Comment by martiniil** [comment]
@jschleicher I like your first suggested change more.

File: prbt_default_application/launch/my_application.launch [hunk 3]
```diff
@@ -59,9 +59,15 @@ limitations under the License.
 
       <!-- Safe-torque-off-input from the safety controller is read to brake the robot before power-off
            The signal is read via modbus and you can configure the modbus server ip -->
-      <!-- <arg name="sto" value="pnoz" /> -->
+      <!-- <arg name="sto" value="pss4000" /> -->
       <!-- <arg name="sto_modbus_server_ip" value="192.168.0.10" /> -->
 
+      <!-- You can additionally select the two safety functions below: -->
+      <!-- Set braketest usage -->
```
Comment (jschleicher) [inline prbt_default_application/launch/my_application.launch]: The comment should explain
* When would you want to enable this? Real-hardware only?
* What effect does this have and relation between the features e.g. "read via modbus, so you need to also set the sto_modbus_server_ip above" --> consistent naming??

**Review by jschleicher — CHANGES_REQUESTED** [review]
Doesn't work with the current version :-(

If I uncomment the python script (last line of the launch file) I get the following error:
```
Traceback (most recent call last):
  File "/root/ws/src/pilz_templates/prbt_default_application/scripts/myApplication.py", line 90, in <module>
    r = Robot(__REQUIRED_API_VERSION__)  # instance of the robot
  File "/opt/ros/melodic/lib/python2.7/dist-packages/pilz_robot_programming/robot.py", line 135, in __init__
    self._establish_connections()
  File "/opt/ros/melodic/lib/python2.7/dist-packages/pilz_robot_programming/robot.py", line 524, in _establish_connections
    rospy.wait_for_service(self._GET_SPEED_OVERRIDE_SRV, self._SERVICE_WAIT_TIMEOUT_S)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 145, in wait_for_service
    raise ROSException("timeout exceeded while waiting for service %s"%resolved_name)
rospy.exceptions.ROSException: timeout exceeded while waiting for service /prbt/get_speed_override
```
See PilzDE/pilz_industrial_motion/issues/230

**Comment by jschleicher** [comment]
I've drawn the rename and PEP-8 out of this PR so this PR is just about braketest and opmode as its title suggests.

**Comment by jschleicher** [comment]
@martiniil @JonathanGruner I gave it another try rephrasing and grouping the options. I'm still unsure, what the `sto` option does? Is this used for braketest and operationmode as well? And why should a user disable operation mode support? I see just *one* usecase *with* safety function and *one without* - no need for finegrain feature selection.

**Comment by martiniil** [comment]
> @martiniil @JonathanGruner I gave it another try rephrasing and grouping the options. I'm still unsure, what the `sto` option does? Is this used for braketest and operationmode as well? And why should a user disable operation mode support? I see just _one_ usecase _with_ safety function and _one without_ - no need for finegrain feature selection.

@jschleicher Yes, I think the use-cases you desribe are sufficient. Currently, `sto` serves for specifying the safety controller, which is of course misleading.

edit: The braketest-module may be missing, but this could also be decided dependent on the name of the safety controller?

**Comment by ct2034** [comment]
I would suggest the following:
We merge this, as it is now. Because this will work with the current state of `pilz_robots`. 

And then we continue the discussion about naming of sto there. There are more aspects to this: https://github.com/PilzDE/pilz_robots/pull/315
Ok, @martiniil @jschleicher ?

**Comment by JonathanGruner** [comment]
I read through your comments, and from mine/customer point of view I would write both tags below the tag with "sto" (which I requested to change to "device" because a STO is something else, see https://github.com/PilzDE/pilz_robots/pull/315 , and than maybe you can define your own device in the future), but commented. 

But I would prefer to discuss how we design the user frontend together in the next sprint review 
on Wednesday.

**Comment by ct2034** [comment]
remaining changes addressed in #16

**Pull request closed without merging**